### DISCUSSION TITLE
0.14.0-stable has been released!

### DISCUSSION THREAD

**Discussion by mihe** [godot-jolt/godot-jolt#1009] [Announcements]
This release only supports **Godot 4.3** (starting with [4.3-stable](https://godotengine.org/releases/4.3/)).

### Changed

- ⚠️ Changed the `center_of_mass` property of `PhysicsDirectBodyState3D` to be a relative global position rather than an absolute global position, to match Godot Physics.
- ⚠️ Changed the ray-cast parameter `hit_back_faces` to no longer hit the back/inside of convex shapes, to better match Godot Physics. This means `hit_back_faces` will only have an effect on `ConcavePolygonShape3D` and `HeightMapShape3D`.
- ⚠️ Changed the ray-cast parameter `hit_back_faces` to only hit back-faces of a `ConcavePolygonShape3D` if its `backface_collision` property is set to `true`, to better match Godot Physics.
- ⚠️ Changed the `get_applied_force` and `get_applied_torque` methods of `JoltHingeJoint3D`, `JoltSliderJoint3D`, `JoltConeTwistJoint3D` and `JoltGeneric6DOFJoint3D` to now also include forces/torque applied by any enabled motor or spring as well.

### Added

- Added support for `WorldBoundaryShape3D`.
- Added support for `face_index` in ray-cast results.
- Added project setting "World Boundary Shape Size", to allow changing the effective size of `WorldBoundaryShape3D`.
- Added project setting "Use Legacy Ray Casting", to allow reverting back to the (now legacy) behavior of the `hit_back_faces` ray-cast parameter.
- Added project setting "Enable Ray Cast Face Index", to allow opting in to the additional memory needed to support `face_index`, which is roughly an additional 25% to every `ConcavePolygonShape3D`.
- Added "Max Force"/"Max Torque" to the spring properties of `JoltGeneric6DOFJoint3D`.

### Fixed

- Fixed issue with `SoftBody3D` not waking up when doing things like changing its `total_mass`, `damping_coefficient`, `simulation_precision`, or when moving any of its pinned points.
- Fixed issue where bodies with multiple shapes could end up clipping through other bodies if the "Use Enhanced Internal Edge Detection" setting was enabled (which it is by default).
- Fixed crash when changing the mesh of an in-tree `SoftBody3D` that has pinned vertices.
- Fixed issue where the joint substitute nodes (`JoltHingeJoint3D`, etc.) would not display their angle properties in degrees when using a `DISABLE_DEPRECATED` build of Godot.
- Fixed crash when a `SoftBody3D` collided with the border of a `HeightMapShape3D` using non-power-of-two dimensions.
- Fixed issue where CCD would have no effect against a `ConcavePolygonShape3D` that had `backface_collision` enabled.
- Fixed crash when using a `SeparationRayShape3D` in a `RigidBody3D` that had CCD enabled.
- Fixed issue where the behavior of the `body_test_motion` method of `PhysicsServer3D` (which powers `move_and_collide` and `move_and_slide`) may have differed from Godot Physics when dealing with a `SeparationRayShape3D` that had `slide_on_slope` enabled.

<hr /><em>This discussion was created from the release <a href='https://github.com/godot-jolt/godot-jolt/releases/tag/v0.14.0-stable'>0.14.0-stable</a>.</em>

**Comment by V1LLz** [comment]
First of all, thank you for the new version, and secondly, is there any plan to eventually add the features available in the `JoltGeneric6DOFJoint3D` node to the implicit joint in `PhysicalBone3D`? Features like motors or the newly added Max Force/Max Torque in this version. Or does this entirely depend on what’s proposed by Godot? Thanks in advance! :D

**Reply by mihe** [reply]
That would unfortunately require changes on the Godot side of things to make happen.

There is [a proposal](https://github.com/godotengine/godot-proposals/issues/10438) for allowing the physics implementation to add custom properties on the various physics scene nodes, like the `PhysicalBone3D` node, which I started working on a few weeks back but haven't had the time/energy to finish.

I plan to pick it back up as part of migrating this extension to Godot itself, as I don't really see any better alternatives for managing some of the Jolt-specific features that ideally should be brought over in the migration.

So basically, this will likely be solved eventually, but not necessarily for this extension per se.
